### PR TITLE
Added delete of value labels to the Prepare > Column: Factor > View Levels/Labels dialogue using delete key in column metadata

### PR DESCRIPTION
**Pull request by N-thony**
Fixes (partially) #7073 
This is still a work in progress.
@dannyparsons I need your help here. Could we have a call? Thanks

### CODE REVIEW

**Comment by N-thony** [comment]
> Fixes (partially) #7073 This is still a work in progress. @dannyparsons I need your help here. Could we have a call? Thanks

@dannyparsons could we have a discussion on this?

**Comment by N-thony** [comment]
> @N-thony that seems fine, as far as it goes, until you have discussed with @dannyparsons Keep asking him! a) I notice that Ok isn't enabled for the data frame, when you first use the delete option. b) If you go to the columns, and choose a variable, and return to the data frame, then it is enabled. But it currently delivers the same command for the columns, not anything for the whole data frame. But that will change anyway. c) Related, if easy, will be to delete a label from the metadata. I notice that this works in principle, but not in practice. It delivers the following command:
> 
> ```
> # Edited variables metadata value
> data_book$append_to_variables_metadata(data_name ="survey",col_names = "fert",property="labels",new_val="")
> ```
> 
> Isn't that the sort of thing you want as your function. I also tried with:
> 
> `data_book$append_to_variables_metadata(data_name ="survey",col_names = "fert",property="labels",new_val="NA") ` Maybe you could be trying that as the code in your dialogue. By the way it didn't seem to work, through the metadata window. Perhaps it will work for you, or that's what you could investigate while waiting for Danny?
> 
> Once sorted, then we should also see what needs to change in the metadata window. Currently it just seems to mess up the factor!
@rdstern thanks for review though this is pending and will be ready for review after the help from Danny. 
@dannyparsons could we discuss this or maybe guide me probably on the backend side? Thanks.

**Comment by N-thony** [comment]
@rdstern @shadrackkibet I have added the delete labels feature on View Levels/Labels dialogue, and in addition I have added the possibility of deleting the labels from `Delete` keyboard. Please take a look. Thanks.

**Review by rdstern — CHANGES_REQUESTED** [review]
@N-thony I don't know if you  checked the results of what you have done with the delete?
I first tried deleting all in the survey data.  Then I found that the levels/Labels dialogue doesn't work at all.  And the Labels has = NA for all variables, while perhaps it should be NA.
When I started again and deleted just the fert (which I had made a labelled factor, Then I was still ok examining the other factors with the Levels/Labels, but not the fert that I had deleted.
When choosing the variables I wonder if the data selector could just have the variables that have value labels to delete?  (Perhaps ideally where there are none, there might be a message saying No Variables have Value Labels to Delete.

**Comment by rdstern** [comment]
I also notice that when there are no labelled variables, then that item is absent in the metadata.  For example, look at the survey data when imported.  There is no Value Label entry.  If you then make fert into a factor, then it has value labels and the entry is provided.  I suggest if you delete value labels from the data frame, then it might delete the entry completely, rather than keeping it with NA in every variable.

**Comment by N-thony** [comment]
> @N-thony I don't know if you checked the results of what you have done with the delete? I first tried deleting all in the survey data. Then I found that the levels/Labels dialogue doesn't work at all. And the Labels has = NA for all variables, while perhaps it should be NA. When I started again and deleted just the fert (which I had made a labelled factor, Then I was still ok examining the other factors with the Levels/Labels, but not the fert that I had deleted. When choosing the variables I wonder if the data selector could just have the variables that have value labels to delete? (Perhaps ideally where there are none, there might be a message saying No Variables have Value Labels to Delete.

@rdstern of course, I did the test and it is very trivial to notice that, I was wondering why it was adding `=` even if you try to delete the label manually on the grid by replacing with nothing or  `NA` it will change it to `= NA`, I had the check on that with @shadrackkibet and I was thinking that the R function need to be improved

image

image

**Comment by N-thony** [comment]
@rdstern @shadrackkibet I have resolved the comment above and made a small change in the used R function here.
@rdstern could you re-test? Thanks.
@shadrackkibet could you also test and check in the R function? Thanks

**Review by rdstern — DISMISSED** [review]
I am approving, but add some items I noticed.
a) Ideally when you choose Selected Variables, the data selector would say Labelled Variables and just give a list of them - rather than all variables.  I don't know if that is easy.  I am hoping that @shadrackkibet Select feature could also select all labelled variables in a data frame?
b) I don't know if the editing of the metadata is in the same pull request.  There I noticed that the Delete button works nicely.  I wouldn't mind if the right-click, when you choose one-or-more cells here also had Delete.  Currently it just has Help.
c) One item (I presume) to turn off, is that you can currently edit within those fields.  There you can really mess up the labelling.  That needs to be turned off.  So, then there is nothing you can do to value labels in that grid, excpet delete them.

File: instat/dlgViewLabelsAndLevels.vb [hunk 1]
```diff
@@ -132,7 +180,38 @@ Public Class dlgViewFactorLabels
         TestOkEnabled()
     End Sub
 
-    Private Sub ucrReceiverFactorColumns_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrReceiverVariables.ControlContentsChanged, ucrChkShowFrequencies.ControlContentsChanged, ucrChkShowLabels.ControlContentsChanged, ucrChkShowPercentage.ControlContentsChanged, ucrChkShowType.ControlContentsChanged, ucrChkShowValues.ControlContentsChanged
+    Private Sub SetReceiverVariableVisible()
+        If rdoDeleteValueLabels.Checked Then
+            ucrReceiverVariables.Visible = If(rdoSelectedColumn.Checked, True, False)
```
Comment (lloyddewit) [inline instat/dlgViewLabelsAndLevels.vb:185]: simplify?
```suggestion
            ucrReceiverVariables.Visible = rdoSelectedColumn.Checked
```

File: instat/ucrColumnMetadata.vb [hunk 2]
```diff
@@ -122,6 +123,26 @@ Public Class ucrColumnMetadata
...
+
+                clsDeleteLabelsFunction.SetRCommand(frmMain.clsRLink.strInstatDataObject & "$append_to_variables_metadata")
+                clsDeleteLabelsFunction.AddParameter("data_name", Chr(34) & _grid.CurrentWorksheet.Name & Chr(34), iPosition:=0)
+                clsDeleteLabelsFunction.AddParameter("col_names", frmMain.clsRLink.GetListAsRString(_grid.GetSelectedColumns), iPosition:=1)
+                clsDeleteLabelsFunction.AddParameter("property", Chr(34) & "labels" & Chr(34), iPosition:=2)
+                clsDeleteLabelsFunction.AddParameter("new_val", Chr(34) & Chr(34), iPosition:=3)
+                frmMain.clsRLink.RunScript(clsDeleteLabelsFunction.ToScript())
+            End If
+        Else
+            MsgBox("Deleting cells is currently disabled. This feature will be included in future versions." & Environment.NewLine & "To remove a cell's value, replace the value with NA.", MsgBoxStyle.Information, "Cannot delete cells.")
```
Comment (lloyddewit) [inline instat/ucrColumnMetadata.vb:142]: add line breaks to make more readable

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@N-thony thanks, just a few questions

**Comment by N-thony** [comment]
> Event DeleteLabels(strColumnName As String)

@rdstern just a typo when pushing, could you test again?

File: instat/Interface/IColumnMetaDataGrid.vb [hunk 3]
```diff
@@ -19,6 +19,8 @@ Public Interface IColumnMetaDataGrid
 
     Event EditValue(iRow As Integer, strColumnName As String, strPreviousValue As String, newValue As Object)
 
+    Event DeleteLabels(strColumnName As String)
```
[the remark this answers is not in the document]
Comment (N-thony) [reply instat/Interface/IColumnMetaDataGrid.vb:22]: This is more specific for the metadata grid, the reason why that is specific.

**Comment by N-thony** [comment]
@lloyddewit @rdstern I resolved and commented your reviews above. Please test/check if you are happy with? Thanks

**Review by rdstern — APPROVED** [review]
I am still happy and relieved to approve this.  If possible I would still like the selector to indicate and just show the Labelled Variables.  But that may come later

File: instat/Interface/IColumnMetaDataGrid.vb (hunk 3 above)
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/Interface/IColumnMetaDataGrid.vb:22]: thanks, makes sense

File: instat/dlgLabelsLevels.vb [hunk 4]
```diff
@@ -114,21 +117,30 @@ Public Class dlgLabelsLevels
         TestOKEnabled()
     End Sub
 
-    Private Sub ucrReceiverLabels_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrReceiverLabels.ControlValueChanged
-        Dim iMissingValue As Integer
-        Dim clsGetColumnFunction As RFunction = ucrReceiverLabels.GetVariables()
+    Private Sub CountLevels()
+        clsGetColumnFunction = ucrReceiverLabels.GetVariables()
```
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/dlgLabelsLevels.vb:122]: @N-thony this still needs resolving

File: instat/dlgLabelsLevels.vb [hunk 5]
```diff
@@ -19,6 +19,8 @@ Public Class dlgLabelsLevels
     Private bFirstLoad As Boolean = True
     Private bReset As Boolean = True
     Private clsViewLabelsFunction, clsSumCountMissingFunction As New RFunction
+    Private iMissingValue As Integer
```
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/dlgLabelsLevels.vb]: @N-thony this still needs resolving

File: instat/dlgLabelsLevels.vb [hunk 6]
```diff
@@ -19,6 +19,8 @@ Public Class dlgLabelsLevels
     Private bFirstLoad As Boolean = True
     Private bReset As Boolean = True
     Private clsViewLabelsFunction, clsSumCountMissingFunction As New RFunction
+    Private iMissingValue As Integer
+    Private clsGetColumnFunction As RFunction
```
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/dlgLabelsLevels.vb:23]: @N-thony this still needs resolving

**Comment by lloyddewit** [comment]
> @lloyddewit @rdstern I resolved and commented your reviews above. Please test/check if you are happy with? Thanks

@N-thony there are many comments above still unresolved. I started going through them and stating '@N-thony this still needs resolving' buth there are too many. Please go through all the comments again (some may be hidden) and ensure that you've pushed all your changes, thanks

**Comment by N-thony** [comment]
> > @lloyddewit @rdstern I resolved and commented your reviews above. Please test/check if you are happy with? Thanks
> 
> @N-thony there are many comments above still unresolved. I started going through them and stating '@N-thony this still needs resolving' buth there are too many. Please go through all the comments again (some may be hidden) and ensure that you've pushed all your changes, thanks

I think all is resolved now.

**Review by lloyddewit — APPROVED** [review]
@N-thony all my comments are now resolved, thanks

**Comment by lloyddewit** [comment]
@rdstern Please could you test again?
@shadrackkibet Would you also like to test/review this PR or may we go ahead and merge?
Thanks

**Comment by lloyddewit** [comment]
> I am still happy and relieved to approve this. If possible I would still like the selector to indicate and just show the Labelled Variables. But that may come later

@N-thony Will you implement this in a separate PR?

**Comment by shadrackkibet** [comment]
@lloyddewit we just had discussions about this. Just a few things to sort out then we can merge.

**Review by shadrackkibet — APPROVED** [review]
@lloyddewit over to you.

**Comment by lloyddewit** [comment]
@rdstern If you're able to retest and approve then we can merge, thanks

**Review by rdstern — APPROVED** [review]
Great

**Pull request merged**